grails-app/controllers/org/pih/warehouse/reporting/ReportController.groovy

```diff
@@ -654,4 +660,73 @@ class ReportController {
         render(view: "showCycleCountReport", model: [rows: rows])
     }
 
+    def showForecastReport = {
+        def origin = Location.get(session.warehouse.id)
+        params.origin = origin.name
+
+        // Export as XLS
+        if (params.format == "text/csv") {
+            params.originId = session.warehouse.id
+            DateFormat dateFormat = new SimpleDateFormat("MM/dd/yyyy")
+            if (params.startDate && params.endDate) {
+                params.startDate = dateFormat.parse(params.startDate)
+                params.endDate = dateFormat.parse(params.endDate)
+            } else {
+                Integer demandPeriod = grailsApplication.config.openboxes.forecasting.demandPeriod ?: 365
+                Date today = new Date()
+                params.startDate = today - demandPeriod
+                params.endDate = today
+            }
+            def data = reportService.getForecastReport(params)
+
+            def rows = []
+            data.groupBy { it.productId }.collect { productId, items ->
+                def inventoryLevel
+                if (!params.replenishmentPeriodDays || !params.leadTimeDays) {
+                    inventoryLevel = InventoryLevel.findByProduct(Product.load(productId))
+                }
+                def replenishmentPeriodDays = params.replenishmentPeriodDays ?: inventoryLevel && inventoryLevel.replenishmentPeriodDays ? inventoryLevel.replenishmentPeriodDays : 365
+                def leadTimeDays = params.leadTimeDays ?: inventoryLevel && inventoryLevel.expectedLeadTimeDays ? inventoryLevel.expectedLeadTimeDays : 365
+                def productExpiry = forecastingService.getProductExpiry(Location.load(params.originId), replenishmentPeriodDays + leadTimeDays, productId)
+                def qtyOnOrder = items.qtyOnOrder.sum() ?: 0
+                def qtyOnHand = items.qtyOnHand.sum() ?: 0
+                def qtyExpiring = productExpiry.collect {it.quantity_on_hand}.sum() ?: 0
+                def qtyAvailable = qtyOnOrder + qtyOnHand - qtyExpiring ?: 0
+                def avgDemand = productExpiry.collect {it.average_daily_demand}.size() > 0 ? productExpiry.collect {it.average_daily_demand}.get(0).setScale(1, RoundingMode.HALF_UP) : 0
```

Comment:
@MagdaToczek Should the average demand be based on the product expiry?

Reply:
@MagdaToczek I dug further into `product_demand_details`, and I think this is ok. It was a bit misleading at first, what was pulled here.